did/ion/request.go

```diff
@@ -0,0 +1,302 @@
+		RecoveryCommitment: recoveryCommitment,
+	}
+
+	return &CreateRequest{
+		Type:       Create,
+		SuffixData: suffixData,
+		Delta:      delta,
+	}, nil
+}
+
+// NewDeactivateRequest creates a new deactivate request https://identity.foundation/sidetree/spec/#deactivate
+func NewDeactivateRequest(didSuffix string, recoveryKey crypto.PublicKeyJWK, signer BTCSignerVerifier) (*DeactivateRequest, error) {
+	// prepare reveal value
+	revealValue, _, err := Commit(recoveryKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// prepare signed data
+	toBeSigned := struct {
+		DIDSuffix   string              `json:"didSuffix"`
+		RecoveryKey crypto.PublicKeyJWK `json:"recoveryKey"`
+	}{
+		DIDSuffix:   didSuffix,
+		RecoveryKey: recoveryKey,
+	}
+	signedJWT, err := signer.SignJWT(toBeSigned)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign JWT: %w", err)
+	}
+	return &DeactivateRequest{
+		Type:        Deactivate,
+		DIDSuffix:   didSuffix,
+		RevealValue: revealValue,
+		SignedData:  signedJWT,
+	}, nil
+}
+
+// NewRecoverRequest creates a new recover request https://identity.foundation/sidetree/spec/#recover
+func NewRecoverRequest(didSuffix string, recoveryKey, nextRecoveryKey, nextUpdateKey crypto.PublicKeyJWK, document Document, signer BTCSignerVerifier) (*RecoverRequest, error) { //revive:disable-line:argument-limit
+	// prepare reveal value
+	revealValue, _, err := Commit(recoveryKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// prepare delta
+	patches := []any{
+		ReplaceAction{
+			Action:   Replace,
+			Document: document,
+		},
+	}
+
+	_, updateCommitment, err := Commit(nextUpdateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	delta := Delta{
+		UpdateCommitment: updateCommitment,
+		Patches:          patches,
+	}
+
+	// prepare signed data
+	deltaCanonical, err := CanonicalizeAny(delta)
+	if err != nil {
+		return nil, err
+	}
+	deltaHash, err := HashEncode(deltaCanonical)
+	if err != nil {
+		return nil, err
+	}
+	_, recoveryCommitment, err := Commit(nextRecoveryKey)
+	if err != nil {
+		return nil, err
+	}
+
+	toBeSigned := struct {
+		RecoveryCommitment string              `json:"recoveryCommitment"`
+		RecoveryKey        crypto.PublicKeyJWK `json:"recoveryKey"`
+		DeltaHash          string              `json:"deltaHash"`
+	}{
+		RecoveryCommitment: recoveryCommitment,
+		RecoveryKey:        recoveryKey,
+		DeltaHash:          deltaHash,
+	}
+	signedJWT, err := signer.SignJWT(toBeSigned)
+	if err != nil {
+		return nil, err
+	}
+	return &RecoverRequest{
+		Type:        Recover,
+		DIDSuffix:   didSuffix,
+		RevealValue: revealValue,
+		Delta:       delta,
+		SignedData:  signedJWT,
+	}, nil
+}
+
+type StateChange struct {
+	ServicesToAdd        []Service
+	ServiceIDsToRemove   []string
+	PublicKeysToAdd      []PublicKey
+	PublicKeyIDsToRemove []string
+}
+
+func (s StateChange) IsValid() error {
+	// check if services are valid
+	// build index of services to make sure IDs are unique
+	services := make(map[string]Service)
```

Comment:
No need to map to actual services. Also, hint at the right size. 
```suggestion
	services := make(map[string]struct{}, len(s.ServicesToAdd))
```